pkg/utils/kubernetes/networking/services.go

```diff
@@ -54,66 +55,72 @@ func WaitForServiceLoadBalancerAddress(ctx context.Context, c *kubernetes.Client
+				correspondingSvcNameLabel: name,
 			},
 		},
 		Spec: corev1.ServiceSpec{
 			Type: corev1.ServiceTypeLoadBalancer,
-			// copy the selector and ports of the service to check.
+			// Copy the selector and ports of the service to check.
 			Selector: service.Spec.Selector,
 			Ports:    service.Spec.Ports,
 		},
 	}
 
-	// empty selector, we should create the endpoints separately.
-	// if the target service does not have a selector, it usually means that
+	// Empty selector, we should create the endpoints separately.
+	// If the target service does not have a selector, it usually means that
 	// the endpoints of the target server is manually created, but not chosen from pods by labels in selector.
 	// so we need to manually create the same endpoints as the target service has here.
-	if service.Spec.Selector == nil || len(service.Spec.Selector) == 0 {
-		endpoints, err := c.CoreV1().Endpoints(namespace).Get(ctx, name, metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-		// copy the endpoints for the lb service, and set metadata.
-		tempEndpoints := endpoints.DeepCopy()
-		tempEndpoints.ObjectMeta = metav1.ObjectMeta{
-			Namespace: namespace,
-			Name:      lbServiceName,
-			Labels: map[string]string{
-				"corresponding-service": name,
-			},
-		}
-		_, err = c.CoreV1().Endpoints(namespace).Create(ctx, tempEndpoints, metav1.CreateOptions{})
+	if len(service.Spec.Selector) == 0 {
+		endpointSlices, err := c.DiscoveryV1().EndpointSlices(namespace).List(
+			ctx, metav1.ListOptions{LabelSelector: discoveryv1.LabelServiceName + "=" + name},
+		)
 		if err != nil {
 			return err
 		}
 
-		defer func() {
-			err := c.CoreV1().Endpoints(namespace).Delete(ctx, lbServiceName, metav1.DeleteOptions{})
-			if err != nil && !errors.IsNotFound(err) {
-				fmt.Printf("failed to delete endpoints %s/%s after testing, error %v\n",
-					namespace, lbServiceName, err)
+		// Recreate EndpointSlices for the lb service with proper metadata.
+		tempEndpointSlices := endpointSlices.DeepCopy().Items
+		for i := range tempEndpointSlices {
+			epsName := fmt.Sprintf("%s-%d", lbServiceName, i)
+			tempEndpointSlices[i].ObjectMeta = metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      epsName,
+				Labels: map[string]string{
+					discoveryv1.LabelServiceName: lbServiceName, // Maps EndpointSlice to Service.
+					correspondingSvcNameLabel:    name,
+				},
+			}
+			if _, err = c.DiscoveryV1().EndpointSlices(namespace).Create(ctx, &tempEndpointSlices[i], metav1.CreateOptions{}); err != nil {
+				return err
 			}
-		}()
+			// For each successfully created temporary EndpointSlice ensure deletion on return from the function.
+			defer func(namespace, epsName string) {
+				err := c.DiscoveryV1().EndpointSlices(namespace).Delete(ctx, epsName, metav1.DeleteOptions{})
+				if err != nil && !errors.IsNotFound(err) {
+					fmt.Printf("failed to delete EndpointSlice %s/%s after testing, error %v\n",
+						namespace, epsName, err,
+					)
```

Comment:
This might be refactored into returning this error in the enclosing `func` so that it's not lost among the logged messages.

We might create an issue for this for later, to solve it in a separate PR.

Reply:
Yeah, let's create an issue. I think the whole KTF should have some more standardised logging / error returning strategy

Reply:
While at it: do you mind creating an issue to test `WaitForConnectionOnServicePort()`? We don't want to rely on KIC's tests to inform us about failures in its implementation.